Rapidhash V3 implementation

hoxxep opened the pull request:
@Nicoshev found some further performance to squeeze out of rapidhash. This takes a lot of the learnings from the rust implementation. Happy to jump on a call sometime this week if you'd like to discuss it?

- Use a randomised seed, the same as V1.
- Don't re-mix the seed — for random and well-chosen fixed seeds this is wasted effort. For sequential seeds we provide a separate `rapid_preseed` function to mix the seed. When the function can be fully inlined this might get optimised out, but is helpful for short inputs with run-time selected seeds.
- 4-way branch at inputs:
  - `<=16` rapidhash V2 logic
  - `<=64` rapidhash V2 logic, lightly modified to accept an extra 16 bytes
  - `<=288` rapidhash V1 logic
  - `>288` rapidhash V2 logic as a non-inlined cold path, this avoids clobbering registers on shorter inputs.
- `RAPID_NO_AVALANCHE` minor optimisation to provide an extra single-digit percentage speed improvement by skipping the final avalanche step when an SMHasher grade hash function isn't required, but we want rapidhash speed.

Benchmark results soundly match or beat rapidhash V1 and rapidhash V2 at all input sizes.

<img width="1161" alt="image" src="https://github.com/user-attachments/assets/23a7fb8d-b24e-4941-a634-4dafca5432c6" />

Small request to hold off publishing a V3 until I can confirm the rust implementation sees the same improvements (to compete with foldhash). Thank you!

## TODO
- [ ] Check the noinline attribute macro to include support for other compilers.
- [ ] Test V3 against SMHasher.
- [ ] Confirm V3 is optimal for the rust library.

hoxxep commented on rapidhash.h:
```diff
@@ -73,7 +73,13 @@
- 
+
+ #ifdef __has_attribute(no_inline)
+ # define RAPIDHASH_NO_INLINE __attribute__((noinline))
```
I'm less familiar with portable C and C++, so let me know if this could/should be improved to support more compilers. I think visual studio uses `__declspec(noinline)` for example?

hoxxep commented on rapidhash.h:
```diff
@@ -226,20 +249,112 @@
+ */
+RAPIDHASH_INLINE_CONSTEXPR uint64_t rapid_preseed(const uint64_t seed) RAPIDHASH_NOEXCEPT {
+  return rapid_mix(seed ^ rapid_secret[2], rapid_secret[1]);
+}
```
This is the pre-seeding that has been removed from the hotpath, and is instead presented to the user as a separate function should they need it. A well chosen fixed, random, or default seed shouldn't need this and it can't always be optimised out -- for example is runtime random seeds.

Ups the chances of user error using seeds 0, 1, 2 but makes a 10% difference on small inputs. Are you happy with this from a user experience?

hoxxep commented on rapidhash.h:
```diff
@@ -257,82 +372,58 @@ RAPIDHASH_INLINE_CONSTEXPR uint64_t rapidhash_internal(const void *key, size_t l
+    return rapid_mix(a ^ secret[7] ^ len, b ^ secret[1]);
+  #else
+    return rapid_mix(a ^ len, b);
+  #endif
```
When not avalanching, we still XOR the length. Perhaps it's unnecessary as it's already included in the seed?

hoxxep replied:
This is where we `^ seed` instead of `^ len` on the understanding that the seed has already used the length. In rust, this final avalanching step is done in a separate method that isn't always inlined alongside the rest of the code, and so using the `seed` removes the need to store and update a long-lived length variable.

Nicoshev commented:
Hey @hoxxep, thanks a lot for sharing the proposal : )

This way of structuring the code is likely to impact peak throughput, usually reached beyond 4kb.
The initial seed mixing is marked as constexpr, it will be computed at compile time whenever a constexpr seed is provided - No need to remove it.
I'd suggest to check the produced asm output on Godbolt, both for aarch64 and AMD64.

For medium-sized output code change, we would need to re-do all quality measurements.

I've tried approached similar to this one before publishing, all of them reducing peak throughput.
I'll try out this one, alongside other planned changes later this week

Nicoshev commented:
The compiler being used also needs to be considered, usually peak throughput is reached using clang-17+.

Regarding the seed, most applications just use 0 as seed. It is important of keeping quality high on such seed value.

hoxxep commented:
I haven't been able to benchmark any slowdown in peak throughput between V2 and V3 on an M1 Max with Apple clang version 17.0.0 as the compiler. It's really the same V2 logic but not inlined.

The function call overhead should only be paid once and makes the rest of the logic much easier to inline for smaller inputs as it uses fewer registers, which was an issue I was running into on the rust crate.

---

As for the seed, that's user error in some ways, and if the user's compiler hasn't compiled it away (such as in my benchmarks) then they're inadvertently paying the seed mixing setup cost on every hash. I had replaced the default seed with the V1 seed in the hope most users use the default. 

We could instead xor a single secret which would reduce the folded multiply overhead? So `seed ^= secret[0] ^ len` to protect against the 0 case?

The rust library uses a separate "hash builder" that caches the seed, so it's not an issue there, but it would still be nice to make that instantiation effectively free.

Nicoshev commented:
@hoxxep RAPIDHASH_COMPACT can be added for better small input speed.

Regarding peak throughput, try out measuring with the program within the bench folder

hoxxep commented:
Rapidhash V2.1 (~45k)
```
 ===  benchmarking 1 hash functions  === 
benchmarking large inputs : from 134217728 bytes (log27) to 128 MB (log27) 
rapidhash, 45342
```

Rapidhash V3 (~45k)
```
 ===  benchmarking 1 hash functions  === 
benchmarking large inputs : from 134217728 bytes (log27) to 128 MB (log27) 
rapidhash, 45476
```

Noise was higher between benchmark runs than the difference (I reckon 1stdev is 500 points). Criterion (rust benchmarking framework) can't detect a significant difference between the two implementations for large inputs, and it detects outliers and calculates the mean across 100 samples.

The blue line (rapidhash V3) in my original comment shows the throughput is significantly larger at all inputs <=256 without sacrificing larger inputs. The yellow line was the no-avalanche option.

Nicoshev commented:
@hoxxep Tested it, you were right, there is no reduction in max speed. 

The version you are proposing seems to be faster for sizes 65 - 111, and slower for 112-288 (or around there).

Nicoshev commented:
@hoxxep V2.1 seems to be slightly faster for sizes 17-64

Nicoshev commented:
@hoxxep I'll perform a broader set of experiments later and see what is better

hoxxep commented:
@Nicoshev V2 here is V2.1 (green), and V3 is blue. This benchmark uses rust's criterion framework to take 100 samples in 5s, averages across them, and reports outliers for each input size `[2, 8, 16, 25, 50, 64, 160, 256, 1024, 4096]`. I'm fairly confident the V3 suggestion is stronger (and have spent the past week non-stop on this, digging into assembly to figure out where the bottlenecks were in both the rust and C++ impls 😅). Happy to jump on a call to discuss if you fancy it.

<img width="1169" alt="image" src="https://github.com/user-attachments/assets/f3db4b6d-2c38-49de-8299-3be0b413361d" />

To make it fair I've also omitted the re-seeding from the V2.1 impl (replaced `seed ^= rapid_mix(seed ^ secret[2], secret[1])` with `seed ^= len` and used V1's `RAPID_SEED` as the hardcoded seed) because it doesn't get compiled out across the static library boundary in my benchmarking. It's -22% latency +28% throughput on 64 and 128 bit inputs when it's not optimised out by the compiler. V1 does not have this optimisation in the graphs.

```
hash/rapidhash_cc_v2/str_8                                                                             
                        time:   [1.8916 ns 1.9235 ns 1.9502 ns]
                        thrpt:  [3.8205 GiB/s 3.8734 GiB/s 3.9387 GiB/s]
                 change:
                        time:   [-23.437% -21.459% -19.583%] (p = 0.00 < 0.05)
                        thrpt:  [+24.351% +27.322% +30.612%]
                        Performance has improved.

hash/rapidhash_cc_v2/str_16                                                                             
                        time:   [1.7860 ns 1.8240 ns 1.8576 ns]
                        thrpt:  [8.0219 GiB/s 8.1694 GiB/s 8.3434 GiB/s]
                 change:
                        time:   [-24.488% -22.804% -21.170%] (p = 0.00 < 0.05)
                        thrpt:  [+26.855% +29.541% +32.430%]
                        Performance has improved.
```

Nicoshev commented:
Re-seeding is important, it cannot be omitted. It guarantees high quality across seeds and that seed 0 is not bad.

hoxxep commented:
Could we offer the following API to give the option to avoid reseeding, where `rapidhash_prepared` skips this step and assumes the user has read the manual and already premixed or randomly chosen their seed?
```c
rapidhash(key, len)                          // default seed
rapidhash_seeded(key, len, seed)             // manually seeded
rapidhash_prepared(key, len, premixed_seed)  // manually seeded with a premixed seed
```

Charts with reseeding for completeness, blue is V2.1, green is V3.

![image](https://github.com/user-attachments/assets/aa6726c7-4c6f-4dc3-9c87-3c747cf281e7)

hoxxep commented:
@Nicoshev update on the rust implementation:
- Removed the if <= 64 -- rapidhash V1 was equally quick, so we just use that logic until input length 288 where the cost of setting up and tearing down the 7-lane approach matches the V1 algo.
- Replaced the final `^ len` in the mixing with `^ seed` as the seed is already mixed with the length. I do this because this stops me from tracking the length between hashes for the following reasons:
  - When hashing a struct with 5 variables, I simply add the length to the seed, and don't have to separately store the length.
  - There is no extra instruction imposed on the rust implementation to add/load/store the an extra length variable.
- We still cold path the >288, again to improve the probability of rapidhash and the calling functions from being inlined. In rust it was posing a problem.

This is with the aim of:
- Matching the best performance of V1 and/or V2.2 at all input sizes.
- Improving the probability that rapidhash (and surrounding methods imposed by rust traits) get inlined, which speeds up the rust implementation.
- Allowing the native rapidhash to be small enough that we can fit a "sponge" into the rust implementation -- a rust-specific hack to work around the `Hasher` trait.

I'm probably going to take the decision to implement `rapidhash_v1`, `rapidhash_v2` etc. functions to hash standalone byte streams, and keep these true to the C++ code. However, the rust-native RapidHasher type is likely to diverge and follow the proposed logic in this PR because of the significant speed improvements to the rust-specific implementation. 

I'm happy to maintain a C++ version alongside for benchmarking and SMHasher purposes, and to compare implementations to ensure they don't diverge too far. Alternatively I can make fork a new hasher under a different name to avoid confusion. Can explain more and chat about all the options tomorrow on the call! Cheers 🍻

hoxxep commented on rapidhash.h:
```diff
@@ -237,20 +264,135 @@
+  uint64_t see1 = seed, see2 = seed;
+  uint64_t see3 = seed, see4 = seed;
+  uint64_t see5 = seed, see6 = seed;
+#ifdef RAPIDHASH_UNROLLED
```
We could remove these macros with V3, as we have no-inline attributes on the cold path, so these instructions will only be included once in the final binary.

ScottPJones commented:
Does mixing the length into the seed at the beginning help much compared to mixing it in at the end?
I ask, because in Julia, a requirement of the hash function (for the Dictionary type) is that all AbstractString types need to produce the same hash value as the Julia String type (which is based off of UTF-8).
I'd like to be able to calculate as efficiently as possible the rapidhash for Latin 1, UTF-16 and UTF-32 encoded strings.
Mixing it into the seed at the beginning, means that you can't do a conversion / hashing of string on the fly (for example, by filling up a 48 byte buffer, do the hash calculations, and continuing on until the end of the string is reached), you have to either convert the entire string upfront (and allocate a large buffer to hold it), or at the very least, do two passes, one to calculate the length of the converted string, and then another pass to do the hash calculations a chunk at a time.

I'd like to base my translation on the very latest, most performant version of Rapidhash, I suppose v3 (especially if that could move the mixing of the length to the last step).

I also have an issue with the license.  The C++ version is BSD 2-clause, but the Rust version is MIT or Apache 2.0.
Could I get permission to use an MIT license for my Julia version as well?  (or is the Rust version licensed incorrectly, AFAIK, as according to BSD 2, it should keep the original BSD 2 license)

Thanks!

Nicoshev commented:
Hey @ScottPJones, feel free to use the released V2.2

This V3 introduces trade-offs with downsides that are not much noticed on benchmarks, but that could pose as regressions on production workloads. Thus it won't be easy to merge.

ScottPJones commented:
Thanks @Nicoshev!  What about the license issue?  Note that Julia has just changed from using MurmurHash3 to Rapidhash the the string hash function (just merged on Git recently, not released anywhere), I need to make my own port of Rapidhash to support non-UTF-8 encoded strings (but generating the same hash value as if they had been UTF-8 encode... not easy to do and maintain performance!)

Nicoshev commented:
@ScottPJones You can use whatever license you want, I don't mind ^^

I'll actually do some research later to see what's the best one to use.

hoxxep commented:
@ScottPJones mixing in the length towards the end is an interesting idea. Theoretically it's a worse hash, as the seed won't have time to propagate through as many of the mixing steps, but in practice for <=16 byte inputs it only gets two mixing steps anyway. Doing the calculation just before the final two mix steps would make it equivalent on <=16 byte inputs.

```cpp
  seed += len;  // <-- insert here to make sure it's still included in the final two mixing steps

  a ^= secret[1];
  b ^= seed;

  rapid_mum(&a, &b);
  return rapid_mix(a ^ secret[7] ^ seed, b ^ secret[1]);
```

This would be really useful for streaming hashes though, I have an implementation in the rust algo for stream hashing of files, but it queries the file metadata for the length first, but I'd much prefer a true streaming implementation: https://github.com/hoxxep/rapidhash/blob/rapidhash-v2/rapidhash/src/v3/rapid_file.rs

There are also V1 and V2.2 versions of the streaming hash on that branch. I haven't been able to see the regressions @Nicoshev mentions, as the current V2.2 has a regression compared to V1 on some input lengths in my benchmarks, so if you're able to benchmark what works best for you it would be helpful for us!

One thing rust does is randomly seed the hasher at runtime, so every instance of an program is seeded differently. I don't know if Julia does this, but if you're using a random runtime-known seed, omitting the `seed ^= rapid_mix(seed ^ rapid_secret[2], rapid_secret[1]);` reseeding step will give you a small performance bump, just be careful of sequential seeds.

Hopefully that's helpful!

hoxxep commented:
@ScottPJones I've made the tweak to only add the length into the seed after all data has been read and run this change in V3 through SMHasher3. All tests pass, so it should be sound! 👍

hoxxep commented:
@ScottPJones perhaps we should make a separate issue to discuss or an email chain? Each branch computes a different value, yes, so the [17,64] branch result != [65,288] for where they support the same input size.

I've just built a streaming rapidhash V3 implementation in rust if you'd like something to refer to that implements streaming. I assume you would do the UTF-8 conversion inside the `ChunkedStreamReader` struct. Rapidhash works best on a "chunk", and I reckon byte-by-byte would be much slower. Take a look at the intermediate buffer in the ChunkedStreamReader if you could write chunks of UTF-8 into it? (Optimising the buffer size to something like `min(8192, original_str_len * encoding_multiplier)` would need a few tweaks elsewhere, as the code assumes chunk size < buffer size).

ChunkedStreamReader link: https://github.com/hoxxep/rapidhash/blob/f4e36efa7bc505f72514d32829edb17b2641dc4c/rapidhash/src/v3/rapid_file.rs#L42

Happy to go through it in detail in a separate issue, email, or on a call?

hoxxep commented:
Closing this PR after our discussion. 

Key things that would be preferable to keep for the Rust implementation:
- **Most important:** Mixing the length into the seed happens towards the end. A CLI was a specific ask in the Rust repo and this allows us to stream files and stdin. Passed SMHasher3 with this change. 

You can try out the working CLI example:
```shell
cargo install rapidhash --git https://github.com/hoxxep/rapidhash/ --branch rapidhash-v2 --all-features

rapidhash hello.txt
echo "hello" | rapidhash
```

Nice to haves that the rust Hasher trait might deviate with:
- Final step uses seed instead of len: this lets us not track the length when hashing multiple byte sequences before doing a single final avalanche step.
- The length gets added to the seed instead of XOR.

All summarised here with this C++:
```cpp
  seed += len;
  a ^= secret[1];
  b ^= seed;

  rapid_mum(&a, &b);
  return rapid_mix(a ^ secret[7] ^ seed, b ^ secret[1]);
```
